### ISSUE TITLE
Mixing mockFindAll & mockQuery -- Is that an issue?

### ISSUE THREAD

**Issue by fwoelm** [danielspaniel/ember-data-factory-guy#298]
Hi there!

I'm completing [Ember's tutorial](https://guides.emberjs.com/v2.14.0/tutorial/autocomplete-component/) right now and I'm running into trouble. The tutorial is using mirage but I've been reading great things about ember-data-factory-guy, so I thought I would get started with it right away :)

I have used factory guy successfully for my four other acceptance tests but in this one -- where I'm mixing `mockFindAll` and `mockQuery`, I cannot get the test to pass.

~~~javascript
// tests/acceptance/list-rentals-test.js
import { test } from 'qunit';
import { makeList, mockQuery, mockFindAll } from 'ember-data-factory-guy'
import moduleForAcceptance from 'super-rentals/tests/helpers/module-for-acceptance';

// setup + teardown in module-for-acceptance.js
moduleForAcceptance('Acceptance | list rentals');

//...

test('should filter the list of rentals by city.', function (assert) {
  mockFindAll('rental').returns( {models: makeList('rental', 3) });

  visit('/');

  let seattleRental = makeList('rental', 1, {city: "Seattle"} );
  mockQuery('rental').returns( {models: seattleRental } ).withParams( {city: "Seattle"} );

  fillIn('.list-filter input', 'Seattle');

  andThen(function() {
    assert.equal(find('.listing').length, 1, 'should show 1 listing');
    assert.equal(find('.listing .location:contains("Seattle")').length, 1, 'should contain 1 listing with location Seattle');
  });
});

//...
~~~

Here is the controller:
~~~javascript
// app/controllers/rentals.js
import Ember from 'ember';

export default Ember.Controller.extend({
  actions: {
    filterByCity(param) {
      console.log(param);
      if (param !== '') {
        console.log("Querying...");
        return this.store.query('rental', { city: param });
      } else {
        console.log("Finding all...")
        return this.store.findAll('rental');
      }
    }
  }
});
~~~

To make sure that all the right functions are being called, I put some `console.log` in there. Here is the output:

~~~javascript
[factory-guy] MockFindAll(rental) Object { data: Array[3] } // <-- not quite sure what this first one is about...
                                                         // <-- okay, we're visiting index, param is empty
Finding all...                                   // <-- finding all, this is expected
[factory-guy] MockFindAll(rental) Object { data: Array[3] } // <-- seems good!
Seattle                                             // <-- param has successfully been changed through fillIn
Querying...                                         // <-- controller is correctly using store.query()
[factory-guy] MockFindAll(rental) Object { data: Array[3] }  // <-- factory-guy is using MockFindAll? o_O
~~~

And then the test fails because 3 is not 1.

Interestingly enough, I can get factory guy to produce the correct mockQuery if I comment out mockFindAll, like so:

~~~javascript
test('should filter the list of rentals by city.', function (assert) {
  // mockFindAll('rental').returns( {models: makeList('rental', 3) });
  visit('/');

  let seattleRental = makeList('rental', 1, {city: "Seattle"} );
  mockQuery('rental').returns( {models: seattleRental } ).withParams( {city: "Seattle"} );

  //...
  // factory-guy logger: [factory-guy] MockQuery(rental) Object { data: Array[1] } queryParams: Object { city: "Seattle" }
~~~

But then the test fails because `Not found: /api/rentals` -- which makes sense because `store.findAll` is no longer mocked.

I hope this makes sense :) Sorry for not understanding this -- the answer is probably obvious. But I've been searching and trying a lot and cannot figure it out.

**Comment by danielspaniel** [comment]
Makes sense so far @fwoelm. 
 You seem to be doing pretty well with factory guy .. impressive.  

 I suggest we go right for the fun of asyc/await style tests to solve this one. So here is how it goes. 

 In you ember-cli-build.js file do this:
```js

var EmberApp = require('ember-cli/lib/broccoli/ember-app');

module.exports = function(defaults) {
  var app = new EmberApp(defaults, {
    'ember-cli-babel': {    // add THIS
      includePolyfill: true // add THIS
    },                      // add THIS
   ... blah blah 
```

then for your test add a few async /  awaits like so: 

```js
test('should filter the list of rentals by city.',  async function (assert) {
  mockFindAll('rental').returns( {models: makeList('rental', 3) });

  await visit('/');

  let seattleRental = makeList('rental', 1, {city: "Seattle"} );
  mockQuery('rental').returns( {models: seattleRental } ).withParams( {city: "Seattle"} );

  fillIn('.list-filter input', 'Seattle');
  await click('Search')  // <<<<  I am guessing you for got this?? 
  let listings = await find('.listing');

  assert.equal(listings.length, 1, 'should show 1 listing');
  assert.equal(listings.find('.location:contains("Seattle")').length, 1, 'should contain 1 listing with location Seattle');

});
```

Most of the time with ember tests you get burned by having a router doing things asynchronously and your test does not wait for them to finish.  andThen is used to wait, but it is frequently not used often enough and is hard to figure out when to andThen. 
  
  await is way easier to grok.  

  I await the visit
 await the fill in 
  await the find 
 and then check out the page. 

 Let me know if you need any other ideas / suggestions / or this did not work for you ..

**Comment by fwoelm** [comment]
Thank you so much for the fast response and for jumping right in!

> `await click('Search')  // <<<<  I am guessing you for got this??`

The action is actually attached to onChange. And it correctly triggers, otherwise I would not be getting to the line that says `console.log("Querying...");`.

> I suggest we go right for the fun of asyc/await style tests to solve this one

That's great to know! It didn't (seem) to change anything as far as I can tell but will be using this from now on.

> Let me know if you need any other ideas / suggestions / or this did not work for you ..

For some reason, factory guy appears to be responding with the findAll mock when story.query() is executed -- if mockFindAll() is not defined, then it correctly runs query(). Does that make any sense to you why that could be? Is it a problem to mix mockFindAll and mockQuery in the same test?
~~~javascript
    filterByCity(param) {
      console.log(param);
      if (param !== '') {
        console.log("Querying...");
        return this.store.query('rental', { city: param });
        // when my test is run, this is the output I get (after fillIn has happened)
        // Seattle
        // Querying...
        // [factory-guy] MockFindAll(rental) Object { data: Array[3] }    <-- this one should not be happening, it should be Query(rental)
      } else {
        console.log("Finding all...")
        return this.store.findAll('rental');
      }
    }
~~~

Really appreciate your help!

PS: I'm going to put a stripped down but runnable version on Github really quickly!

**Comment by danielspaniel** [comment]
ok .. in that case 

 try

```js
  await fillIn('.list-filter input', 'Seattle');
```
 
 and that helps a ton to have a repo to play with to two 👍 for that idea

**Comment by fwoelm** [comment]
Ooookay, here we go: https://github.com/fwoelm/mixing-find-all-and-query

Guess what... I fixed it haha... But I'm fascinated by it and not quite sure I understand what's going on behind the scenes.

### This test works:

~~~javascript
// THIS TEST WORKS
test('THIS DOES WORK: it should show 1 listing after filling in input', async function(assert) {

  let seattleRental = makeList('rental', 1, {city: "Seattle"} );
  let otherRentals = makeList('rental', 2);

  // mockQuery('rental'); <-- if mockQuery('rental') comes first, it does not work
  mockQuery('rental', {city: "Seattle"}).returns( {models: seattleRental } );
  mockQuery('rental'); // <-- mockQuery('rental') must come after the line above


  await visit('/');

  await fillIn('.list-filter input', 'Seattle');
  await keyEvent('.list-filter input', 'keyup', 69);

  assert.equal(find('.listing').length, 1, 'should show 1 listing');
  assert.equal(find('.listing .location:contains("Seattle")').length, 1, 'should contain 1 listing with location Seattle');

});
~~~

### This test does not work:

~~~javascript
// THIS TEST DOES NOT WORK
test('THIS DOES NOT WORK: it should show 1 listing after filling in input', async function(assert) {
  mockFindAll('rental').returns( {models: makeList('rental', 3) });

  let seattleRental = makeList('rental', 1, {city: "Seattle"} );
  let otherRentals = makeList('rental', 2);
  mockQuery('rental').returns( {models: seattleRental } ).withParams( {city: "Seattle"} );

  await visit('/');

  await fillIn('.list-filter input', 'Seattle');
  await keyEvent('.list-filter input', 'keyup', 69);

  assert.equal(find('.listing').length, 1, 'should show 1 listing');
  assert.equal(find('.listing .location:contains("Seattle")').length, 1, 'should contain 1 listing with location Seattle');

});
~~~


Why does it not work if I mix mockQuery() and mockFindAll()? Why does the order of mockQuery() matter?

**Comment by danielspaniel** [comment]
You are pretty clever to figure that out. It was my next suggestion!
 Well, the reason it works is that mockFindAll and mockQuery have the same url:  /rentals
 they both go to the same route on the server.
 Furthermore mockjax keeps list of urls it is mocking. 

  if you do mockFindAll then mockQuery the list is 

  '/rentals' 
  '/rentals?query=blah'

  if you do mockQuery and then mockFindAll the list is 

  '/rentals?query=blah'
  '/rentals' 

 And mockjax looks for the first one that matches.

 So .. in the route:
   ( assuming the mockFindAll then mockQuery ) 
  when store.findAll('rental') happens that matches mockjax '/rentals'
  when store.query('rental',{'blah'}) happens that matches mockjax '/rentals'  also 
  
  which in my opinion is a bug and should not happen. So thanks for bringing this up!

  Let me noodle around and see if I can fix that ( or at least make it more clear in the doc's if I can't )

**Comment by fwoelm** [comment]
Sweet! That's a very satisfying answer and explains the behavior!

I guess for the same reason, the order of mockQuery matters:

```
mockQuery('rental');
mockQuery('rental', {key: value})
```

leads to

```
/rentals
/rentals?key=value
```

and I want the specific ones first!

What would be super helpful would be a method to get the 'defined' mockjax URLs in their respective order. Something along the lines of `FactoryGuy.getMockedUrls()` -- that would be awesome for debugging these kind of issues... just saying :stuck_out_tongue: 


Really appreciate the thoughtful responses and your willingness to help :pray:

**Comment by danielspaniel** [comment]
You know that is a pretty darn good idea:  =>  ```FactoryGuy.getMockedUrls()```

For a free bottle of fresh organic crickets would you be willing to take a crack at trying to add that to Factory Guy ?  It just sounds like fun to me, and you might enjoy the playing around with mockjax and FactoryGuy yourself.  And most people love the taste of crickets so ....

**Comment by fwoelm** [comment]
I spent the past two hours looking at your code. It's amazingly well written, clear, concise. Absolutely inspiring!

I understand now that all of the mocks extend the same class (MockRequest) and that because adapter.buildURL returns the same URL regardless of whether the requestType is findAll or query, the mocks also respond to the same URL. In fact, when I test it with my app, the parameters do not actually change anything about the URL that is built by getURL() in `addon/mocks/mock-request.js`. They are all 'api/rentals'.

The reason the mocks work out when defined in the correct order is because of the `extraRequestMatches()` check defined in setupHandler() also in `addon/mocks/mock-request.js`.

I do think this behavior - of matching the URL regardless of what params are passed - counters intuition and expectations. **I think it would be better and more beginner-friendly if accepting any/all params was an opt-in feature rather than the default.** That way, the ordering of mocks would become irrelevant.

The fix would not need to be very complicated. I see three things to be done in the code:

1. Set queryParams to an empty object `{}` as the default on mockFindAll(), mockQuery(), etc...:

~~~javascript
// addon/mocks/mock-find-all-request.js

import MockGetRequest from './mock-get-request';

export default class MockFindAllRequest extends MockGetRequest {

  constructor(modelName) {
    super(modelName, 'findAll');
    this.setValidReturnsKeys(['models', 'json', 'ids','headers']);
    this.setResponseJson(this.fixtureBuilder.convertForBuild(modelName, []));
    this.queryParams = {}; // <-- add empty params object per default
  }

}
~~~

2. Change extraRequestMatches() to skip validation of param equivalence only if param is not an object (rather than skipping it when param is an empty object):

~~~javascript
// addon/mocks/mock-get-request.js

  paramsMatch(settings) {
    if (!Ember.$.isEmptyObject(this.someQueryParams)) {
      return isPartOf(settings.data, this.someQueryParams);
    }
   // replace if (!Ember.$.isEmptyObject(this.queryParams)) { with the following line,
   // so that our {} empty params will *not* skip this equivalence check
   if (Ember.typeOf(this.queryParams) === 'object') {
      return isEquivalent(this.queryParams, settings.data);
    }
    return true;
  }
~~~

3. Add a new method `ignoreParams()` that users can chain to their mocks to explicitly allow any params. Calling it will set queryParams to a non-object, such as undefined or null. It will just the skip param equivalence validation (since we declared above that we will only validate if param is an object).

~~~javascript
// addon/mocks/mock-get-request.js

  // whitelist all parameters / ignore all parameters / accept any parameters
  ignoreParams() {
    this.queryParams = null;
    return this;
  }

  withParams(queryParams) {
    // ...
~~~


One question that remains for me is how this might adversely affect current users of ember-data-factory-guy. I *guess* most people would not even notice but then again, I don't know.

**Comment by danielspaniel** [comment]
Let me ponder that one .. seems like a good idea actually. 
 
Luckily there are over 1000 tests in the test suite, so to find out how it affects current users, you can just fork the repo, make the changes you like, run >  ember test   and you will see if you broke stuff. 

 I reckon you will break a few things, but I think it will be very limited and may not actually impact anyone ( depending on the type of breaking tests ) since this is not doing too much to the actually mocking itself.

  If you don't feel like it .. that is fine too .. I can do later on tonight, or latest  might take me till weekend? Not sure.

**Comment by fwoelm** [comment]
> You know that is a pretty darn good idea: => FactoryGuy.getMockedUrls()
> ... would you be willing to take a crack at trying to add that to Factory Guy ?

I was actually in the process of trying and then I realized that the method getUrl() (in addon/mocks/mock-request.js) always returns the same URL regardless of my params :stuck_out_tongue: That's not factory guy, though, that's due to [DS,buildURL](https://www.emberjs.com/api/ember-data/2.14/classes/DS.BuildURLMixin/methods/buildURL) on which factory guy bases its URL generation.

The source code for buildURL() shows that it actually completely ignores query params!

Source: https://github.com/emberjs/data/blob/v2.14.4/addon/-private/adapters/build-url-mixin.js#L46

~~~javascript
  /*
    Builds a URL for a given type and optional ID.
    By default, it pluralizes the type's name (for example, 'post'
    becomes 'posts' and 'person' becomes 'people'). To override the
    pluralization see [pathForType](#method_pathForType).
    If an ID is specified, it adds the ID to the path generated
    for the type, separated by a `/`.
    When called by RESTAdapter.findMany() the `id` and `snapshot` parameters
    will be arrays of ids and snapshots.
    @method buildURL
    @param {String} modelName
    @param {(String|Array|Object)} id single id or array of ids or query
    @param {(DS.Snapshot|Array)} snapshot single snapshot or array of snapshots
    @param {String} requestType
    @param {Object} query object of query parameters to send for query requests.
    @return {String} url
  */
  buildURL(modelName, id, snapshot, requestType, query) {
    switch (requestType) {
      case 'findRecord':
        return this.urlForFindRecord(id, modelName, snapshot);
      case 'findAll':
        return this.urlForFindAll(modelName, snapshot);
      case 'query':
        return this.urlForQuery(query, modelName);
      case 'queryRecord':
        // ... all of these just send back to _buildURL, sometimes providing additional information
        // but since the final _buildURL ignore query params, it's irrelevant what these other functions
        // do (irrelevant in this case where we care about parameters)
      default:
        return this._buildURL(modelName, id);
    }
  },

  /**
    @method _buildURL
    @private
    @param {String} modelName
    @param {String} id
    @return {String} url
  */
  _buildURL(modelName, id) {
    let path;
    let url = [];
    let host = get(this, 'host');
    let prefix = this.urlPrefix();

    if (modelName) {
      path = this.pathForType(modelName);
      if (path) { url.push(path); }
    }

    if (id) { url.push(encodeURIComponent(id)); }
    if (prefix) { url.unshift(prefix); }

    url = url.join('/');
    if (!host && url && url.charAt(0) !== '/') {
      url = '/' + url;
    }

    return url;
}

~~~


But we could still produce useful output for a getMockedUrls() method.

Here is what the output could look like:

~~~
1) 'api/users' - exact match
2) 'api/rentals' - with params {city: 'sf', title: 'my rental'}
3) 'api/posts' - with some params {author: 'gandalf'}
4) 'api/messages' - with any params
5) 'api/sessions/new' - exact match
6 ...
~~~

Or, if we wanted purely strings, we could have it return an array and use the star symbol as an indicator of 'any additional params':

~~~
[0]: 'api/users'
[1]: 'api/rentals?city=sf&title=my+rental'
[2]: 'api/posts?author=gandalf*'
[3]: 'api/messages?*'
[4]: 'api/sessions/new'
~~~

My only issue with that it is misleading in so far as that the url for #1 could also be reversed: 'api/rentals?title=my+rental&city=sf' -- but I guess most people understand that...? :grin: If so, then I'm all for the latter version. Seems more simple and clear.


If you get the chance to work on any of this, go for it! I've got enough things to keep me busy :stuck_out_tongue: Otherwise, I will give it a try tomorrow (I'm in Thailand, so it's time to sleep). First time with an Ember addon, kinda exciting :)

**Comment by danielspaniel** [comment]
Good writeup so far. 
  I am kinda undecided about how to show the url, I like version 1  and 2. I can go for version 2 if we can make it more clear what the * means.  It seems so fuzzy. 

4) 'api/messages' - with any params   

  that one is the tricky one and what I think caused the bug because it could be findAll or query match  

 and showing like 

 [3]: 'api/messages?*'  

  seems to suggest .. this will match that url with params or not so that is sorta clear to me. 

  anyway ... I was fixing another bug tonight in ED repo so I did not get a chance to do too much on this,
 but I can keep chatting about it till I finish that one.

**Comment by fwoelm** [comment]
On second thought, I think it would be better to put a space in between URL and wildcard. That seems easier to read and avoids conflicts when someone uses a star in their parameters!

~~~
[0]: 'api/users'
[1]: 'api/rentals?city=sf&title=my+rental'
[2]: 'api/posts?author=gandalf *'
[3]: 'api/messages? *'
[4]: 'api/sessions/new'
~~~

Another option would be to use `[any params]` instead of `*` - like so:

~~~
[0]: 'api/users'
[1]: 'api/rentals?city=sf&title=my+rental'
[2]: 'api/posts?author=gandalf [any params]'
[3]: 'api/messages [any params]'
[4]: 'api/sessions/new'
~~~

As for users not understanding what it means -- I guess one would only find about this method by reading about it in the README and so that's where we should also add the explanation about how to understand the star (or whatever syntax we end up using).

PS: I think something went missing here:

> and showing like
> seems to suggest .. this will match that url with params or not so that is sorta clear to me.

**Comment by fwoelm** [comment]
Here is a practical question:

At the moment, factory guy does not keep track of the mocks have been constructed, is that correct? It just delegates all that work to mockjax. Do you know if/how we can access those mocks without having to keep an internal reference?

**Comment by danielspaniel** [comment]
I was pointing out 
[3]: 'api/messages?*'   was alittle unclear and as you suggest 
[3]: 'api/messages [any params]'  makes it a touch clearer 

 FactoryGuy never had a reason to track the mocks before, and I was always relying on mockjax. 
 I think it would work to keep internal list
 because you can clear the list when mockTeardown is called.
 If mockTeardown is not called then your mockjax setup is still in effect, so all the mocks are still valid, so our output would be correct.

**Comment by fwoelm** [comment]
Yeah, if we can avoid keeping our own internal list, that would be even better :)

I haven't had a chance to work on this further. But the next step is to dive into the mockjax code and see if we can somehow access the defined mocks there. They must be stored somewhere...

**Comment by danielspaniel** [comment]
They are .. I think they use the variable mockHandlers  .. not hard to find ( it is not a big js file )

**Comment by danielspaniel** [comment]
I came up with a simpler fix for this one .. thank again for taking the time to bring it up and point out a solution.  My solution was not exactly the same but your ideas helped me to see it. 
This issue is now fixed in the new version =>  2.13.4

 If for any reason you see it not fixed open this up again.

**Comment by fwoelm** [comment]
It's working really well! Thank you so much!
